src/chunk.c

```diff
@@ -1081,42 +1081,63 @@ ts_chunk_find_or_create_without_cuts(Hypertable *ht, Hypercube *hc, const char *
 	ChunkStub *stub;
 	Chunk *chunk = NULL;
 
-	LockRelationOid(ht->main_table_relid, ShareUpdateExclusiveLock);
+	DEBUG_WAITPOINT("find_or_create_chunk_start");
 
 	stub = chunk_collides(ht, hc);
 
 	if (NULL == stub)
 	{
-		ScanTupLock tuplock = {
-			.lockmode = LockTupleKeyShare,
-			.waitpolicy = LockWaitBlock,
-		};
+		/* Serialize chunk creation around the root hypertable */
+		LockRelationOid(ht->main_table_relid, ShareUpdateExclusiveLock);
 
-		/* Lock all slices that already exist to ensure they remain when we
-		 * commit since we won't create those slices ourselves. */
-		ts_hypercube_find_existing_slices(hc, &tuplock);
-		chunk = chunk_create_from_hypercube_after_lock(ht, hc, schema_name, table_name, NULL);
+		/* Check again after lock */
+		stub = chunk_collides(ht, hc);
 
-		if (NULL != created)
-			*created = true;
-	}
-	else
-	{
-		/* We can only use an existing chunk if it has identical dimensional
-		 * constraints. Otherwise, throw an error */
-		if (!ts_hypercube_equal(stub->cube, hc))
-			ereport(ERROR,
-					(errcode(ERRCODE_TS_CHUNK_COLLISION),
-					 errmsg("chunk creation failed due to collision")));
+		if (NULL == stub)
+		{
+			ScanTupLock tuplock = {
+				.lockmode = LockTupleKeyShare,
+				.waitpolicy = LockWaitBlock,
+			};
+
+			/* Lock all slices that already exist to ensure they remain when we
+			 * commit since we won't create those slices ourselves. */
+			ts_hypercube_find_existing_slices(hc, &tuplock);
+
+			chunk = chunk_create_from_hypercube_after_lock(ht, hc, schema_name, table_name, NULL);
+
+			if (NULL != created)
+				*created = true;
 
-		/* chunk_collides only returned a stub, so we need to lookup the full
-		 * chunk. */
-		chunk = ts_chunk_get_by_id(stub->id, true);
+			ASSERT_IS_VALID_CHUNK(chunk);
+
+			DEBUG_WAITPOINT("find_or_create_chunk_created");
+
+			return chunk;
```

Comment:
Unlock will not be called.

Reply:
I guess it also means that there is no test to check lock in this case.

Reply:
I've seen that our general pattern in the code is to return the lock once the transactions ends. However, in cases where we take the lock to create an object and it exists shouldn't it be better to release the lock immediately to reduce the lock contention?

Reply:
We should not unlock when we have created a chunk (this is the reason we took the lock). The lock is released at the end of the transaction when the chunk will become visible to other transactions. There is a test case that shows lock handling for this case; i.e., the process creating the chunk holds a lock here and the other process waits on LockRelationOid above. When the creating transaction completes, the non-creating transaction stops at the "found" waitpoint and at that point no lock is held by either transaction.

Reply:
Do I understand correctly that the transaction might be long, e.g., copying large volumes of data, and thus other threads will need to wait for long time? I guess it might be fine, since other transactions will conflict on writing to the same chunk, right?

Reply:
@k-rus We take the ShareUpdateExclusive lock every time we create a chunk. Process do not block on inserting data. Sometimes the lock is taken even if no chunk was in fact created and this optimizes that case by releasing the lock early when no chunk was created.

So, to summarize, processes only block each other when they create chunks. This is to avoid deadlocks and other issues. For instance, we don't want one process to create chunk A B and another create B A, which would lead to a deadlock. That is why we use this "global" lock on the root hypertable. This locking behavior doesn't change with this PR.

Reply:
I think my question was stupid. There is no reason to unlock, since the created chunk will be visible only after the transaction commit.

Reply:
No stupid questions :)